pkg/graph/resolvers/plan_task_list_section_locks.go

```diff
@@ -0,0 +1,165 @@
+
+	return subscriber.GetChannel(), nil
+}
+
+// LockTaskListSection will lock the provided task list section on the provided model
+func (p PlanTaskListSectionLocksResolverImplementation) LockTaskListSection(ps pubsub.PubSub, modelPlanID uuid.UUID, section string, principal string) {
+	if isSectionLocked(modelPlanID, section) {
+		return
+	}
+
+	status := model.TaskListSectionLockStatus{
+		Section:  section,
+		LockedBy: principal,
+	}
+
+	_, found := taskListSectionLocks[modelPlanID]
+	if !found {
+		taskListSectionLocks[modelPlanID] = make(map[string]model.TaskListSectionLockStatus)
+	}
+
+	taskListSectionLockMutex.Lock()
+	taskListSectionLocks[modelPlanID][section] = status
+	taskListSectionLockMutex.Unlock()
+
+	ps.Publish(modelPlanID, pubsubevents.TaskListSectionLocksChanged, model.TaskListSectionLockStatusChanged{
+		ChangeType: model.ChangeTypeAdded,
+		LockStatus: &status,
+	})
+}
+
+// UnlockTaskListSection will unlock the provided task list section on the provided model
+//	This method will fail if the provided principal is not the person who locked the task list section
+func (p PlanTaskListSectionLocksResolverImplementation) UnlockTaskListSection(ps pubsub.PubSub, modelPlanID uuid.UUID, section string, principal string) {
+	if !isSectionLocked(modelPlanID, section) {
+		return
+	}
+
+	status := taskListSectionLocks[modelPlanID][section]
+	if !isUserAuthorized(status, principal) {
+		return
+	}
+
+	internalUnlockTaskListSection(ps, modelPlanID, section, status)
+}
+
+func internalUnlockTaskListSection(ps pubsub.PubSub, modelPlanID uuid.UUID, section string, status model.TaskListSectionLockStatus) {
+	taskListSectionLockMutex.Lock()
+	delete(taskListSectionLocks[modelPlanID], section)
+	taskListSectionLockMutex.Unlock()
+
+	ps.Publish(modelPlanID, pubsubevents.TaskListSectionLocksChanged, model.TaskListSectionLockStatusChanged{
+		ChangeType: model.ChangeTypeRemoved,
+		LockStatus: &status,
+	})
+}
+
+func isUserAuthorized(status model.TaskListSectionLockStatus, principal string) bool {
+	return principal == status.LockedBy
+}
+
+// UnlockAllTaskListSections will unlock all task list sections on the provided model
+func (p PlanTaskListSectionLocksResolverImplementation) UnlockAllTaskListSections(ps pubsub.PubSub, modelPlanID uuid.UUID) {
+	for section, status := range taskListSectionLocks[modelPlanID] {
+		internalUnlockTaskListSection(ps, modelPlanID, section, status)
+	}
+
+	delete(taskListSectionLocks, modelPlanID)
+}
+
+func isSectionLocked(modelPlanID uuid.UUID, section string) bool {
+	session, found := taskListSectionLocks[modelPlanID]
+	if !found {
+		return false
+	}
+
+	_, found = session[section]
+	return found
+}
+
+// GetTaskListSectionLocks is a convenience relay method to call the corresponding method on a resolver implementation
+func GetTaskListSectionLocks(modelPlanID uuid.UUID) []*model.TaskListSectionLockStatus {
+	return NewPlanTaskListSectionLocksResolverImplementation().GetTaskListSectionLocks(modelPlanID)
```

Comment:
Can you speak to why you chose to use the implementation method here? I feel like it might be clearer to just to duplicate making the object instance here. It's nice that this can be called in one line, but I think it would be clearer to me without the function call. (It feels a lot more complex)

Reply:
There are three main things to address here:

- Why have I made these wrapper functions which simply create and use an object?
-- In order to provide a similar retroactive structure compliant with our other resolvers
- Ok sure, but why create an object? Why not just do the same as we have been doing?
-- Mostly to facilitate the generation of a mock resolver, see pkg/graph/resolvers/mockplantasklistresolver/resolver.go
- What value does the mock resolver provide?
-- As it is, not too much. Currently we can use the mock resolver to satisfy system level testing. However, by building out a system where the server is provided with a mapping of resolvers we greatly improve the flexibility and testability of our servers. That is outside the scope of this PR, having enough side complexity as-is.

TL;DR:
Upon the development of system where GraphQL can access provided resolver instances, these convenience/wrapper functions will be obsolete and should be removed.